### PR TITLE
fix: skip auto-fix no-jinja-when on string-embedded jinja

### PR DESCRIPTION
**Pull request by f1047**
### Summary

The `no-jinja-when` rule's auto-fix stripped out a necessary `{{ ... }}` notation in a string expression.

This PR prevents Jinja2 templates embedded in a quoted string from being transformed (e.g., `"{{ v }}"`, `"version={{ expected_version }}"`).

We need a more sophisticated way to auto-fix an embedded pattern.

### Changes

- Skip auto-fix when a template is in a quoted string.
- Add examples that do not deserve transformation with the current implementation.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved automatic cleanup of Jinja delimiters in `when` expressions.
  * Preserved Jinja syntax when it appears inside quoted strings, including strings with escaped quotes.
  * Preserved expressions where the entire quoted value is a Jinja block.
  * Ensured fixes are reported only when an actual change is made.
* **Tests**
  * Added playbook examples and coverage for quoted Jinja expressions.
  * Updated transformation expectations to reflect the expanded coverage.

### CODE REVIEW

**Comment by rockygeekz** [comment]
@f1047 Thanks for the PR!  Can you solve the merge conflicts in order to proceed with this PR?

File: src/ansiblelint/rules/no_jinja_when.py [hunk 1]
```diff
@@ -17,9 +17,19 @@
...
 _WHEN_KEYS = ("when", "changed_when", "failed_when")
 
 
+def _strip_redundant_jinja(value: str) -> str:
+    for quoted in RE_QUOTED_STRING.findall(value):
+        inner = quoted[1:-1]
+        if "{{" in inner:
+            # {{ }} embedded in a quoted string changes meaning if stripped; skip it
+            return value
+    return RE_JINJA.sub(r"\1", value)
```
[the remark this answers is not in the document]
Comment (f1047) [reply src/ansiblelint/rules/no_jinja_when.py]: Addressed: a6876c4d7998ea5479e16b9c321fe82cdaa03508

**Comment by f1047** [comment]
@rockygeekz Thank you for checking. I resolved the conflicts.

**Review by shvenkat-rh — COMMENTED** [review]
## Review

### `_strip_redundant_jinja` — logic is correct but conservative

The core of the fix:

```python
def _strip_redundant_jinja(value: str) -> str:
    for quoted in RE_QUOTED_STRING.findall(value):
        inner = quoted[1:-1]
        if "{{" in inner:
            return value   # skip the whole expression
    return RE_JINJA.sub(r"\1", value)
```

If **any** quoted string in the expression contains `{{ }}`, the entire expression is skipped. This means an expression like `"{{ safe }}" and "'{{ v }}' == '8'"` won't have `{{ safe }}` stripped either, even though that's safe. The PR description already acknowledges this ("We need a more sophisticated way to auto-fix an embedded pattern"), so this conservative approach is intentional and safe — it never corrupts code, it just defers some transforms.

Worth a comment in the function noting the conservative scope, so future contributors don't think it's a gap to close naively.

### `RE_QUOTED_STRING` regex — correct

`r'"(?:\\.|[^"\\])*"|\'(?:\\.|[^\'\\])*\''` correctly handles escaped delimiters inside strings (`\"`, `\'`), as verified by `test_strip_redundant_jinja_handles_escaped_quotes`. The `ansible_facts["os_family"]` case confirms double-quoted dict access inside `{{ }}` is not mistaken for a quoted string boundary.

### `match.fixed` bug fix — good catch

Previously `match.fixed = True` was set unconditionally, even when `_transform_when_keys` made no change. The refactor to return `bool` from `_transform_when_keys` / `transform_for_roles` and assign `match.fixed = _transform_when_keys(...)` is the right fix.

### In-place list mutation — intentional and correct

The comment in `_transform_when_value` explains why lists are mutated in place rather than rebuilt: ruamel.yaml preserves comments and blank lines when you mutate existing objects. The old list-comprehension approach would silently drop those. This is a meaningful correctness improvement.

### Test coverage — solid

- `test_strip_redundant_jinja_skips_quoted_templates` covers the main cases.
- `test_strip_redundant_jinja_handles_escaped_quotes` covers the escaped-quote edge case.
- Example playbooks and `.transformed.yml` are consistent: cases 4 and 5 are identical in both (no-op transform), which is the expected behavior.
- `test_transformer.py` count bump from 3 → 5 is correct.

### One nit

`transform_for_roles` does `role[key] = new_value` even for list values, where `_transform_when_value` already mutated the list in place and returns the same object. Assigning it back is harmless but slightly misleading. Could guard with `if not isinstance(new_value, list):` or just leave it — not a correctness issue.

Overall the fix is safe and well-tested. The conservative skip scope is the right call given the complexity of properly handling partial expressions.

**Comment by shvenkat-rh** [comment]
@f1047 please rebase it. The PR looks good to me

**Review by shvenkat-rh — APPROVED** [review]
